### ISSUE TITLE
Add ability to consume Step Templates in Deployment Processes

### ISSUE THREAD

**Issue by Shanjohn** [OctopusDeployLabs/terraform-provider-octopusdeploy#137]
Is there a plan to support usage of step templates in octopusdeploy_deployment_process in the future?

**Comment by ray-jam** [comment]
I managed to get something working via a standard step action:

```
  step {
    name      = "Terraform - PlanV2"
    condition = "Success"
    start_trigger = "StartAfterPrevious"
    action {
      action_type                        = "Octopus.TerraformPlan"
      can_be_used_for_project_versioning = true
      is_disabled = false
      is_required = true
      name        = "Terraform - PlanV2"
      properties = {
        "Octopus.Action.Package.DownloadOnTentacle"             = "False"
        "Octopus.Action.Package.FeedId"                         = "feeds-builtin"
        "Octopus.Action.Package.PackageId"                      = "#{Octopus.Project.Name}"
        "Octopus.Action.Script.ScriptSource"                    = "Package"
        "Octopus.Action.Template.Id"                            = "ActionTemplates-123"
        "Octopus.Action.Template.Version"                       = "3"

```
Found it useful to import from existing projects to get a working solution:
- Create a sample project in OD
- add the step processes you want
- create a local terraform project and add a `octopusdeploy_deployment_process ` resource, don't worry about the step details
- run a `terraform import <RESOURCE-ADDRESS> <OD-ID>` 
- run a plan and it will show the differences which is effectively the data you want to grab or you can go through the statefile for details

From that information you can build you terraform code to suit the step template and clear out your temporary bits. May not be the long term best solution but for now it means we can proceed with deploying our OD build processes using step templates.

**Comment by Shanjohn** [comment]
@jbristowe It is completely ok for usual steps, but what about step templates ? When you use it, the right solution seems to be just to pass step template parameters as we do in GUI, but here we need to basically copy all of internal step template insides via properties manually

**Comment by blytheaw** [comment]
I'm unclear on if using library step templates within the deployment process is actually supported. I raised issue #184 because I can't seem to import a deployment process that uses a library step template.

Additionally, I can't seem to create a new deployment resource using a library step template without specifying all the properties of the library step template directly, which defeats the purpose, as @Shanjohn mentioned above. 

@jbristowe can you provide some clarity on if this pattern is supported for the octopusdeploy_deployment_process resource?

**Comment by angusmcdonald** [comment]
The properties can be attached by including add them in the **properties** object in the **deploy_package_action** 

I just raised a ticket #196 (which most likely can be linked to this)  as this object has been flagged to be deprecated and its seems like a workaround  

an example of this workaround is in there

**Comment by chathsuom** [comment]
I am also facing the same issue. We can't use step template without defining the content of the step template.

**Comment by waylew50** [comment]
any update on this. I also get this warning in v0.8.1 "Warning: Argument is deprecated
│
│   with module.project.octopusdeploy_deployment_process.main,
│   on ..\..\modules\project\process.tf line 24, in resource "octopusdeploy_deployment_process" "main":
│   24:       properties = {

seems like the work around for using properties to reference step templates. may be impacted by this in a coming release? hopefully, there are plans for a step_template resources and data blocks?

**Comment by lee5i3** [comment]
I'm really hoping for progress on this, being able to keep step templates in sync across spaces would be a big help!

We currently have around 100 deployments which are identical over 5 spaces, would be nice to be able to keep the same deployment processes the same.

**Comment by Clare-Octopus** [comment]
Another user seeing this issue (internal) - https://octopus.zendesk.com/agent/tickets/172218

**Comment by haraldsk** [comment]
Just want to chime in that this is an issue for us as well, and working around it makes our code become really messy. It would be great if it got some traction.

**Comment by mjhilton** [comment]
Agreed that though you can technically use a Step Template in a Deployment Process via Terraform, doing so basically negates the benefits that step templates bring. It's very messy and involves "copying" the logic of the Step Template into the properties of the action in the process.

**Comment by KennethBates** [comment]
Another user affected (internal): https://octopus.zendesk.com/agent/tickets/225624

**Comment by mjhilton** [comment]
To provide an update on this: we understand the pain and limitation that this causes for using Terraform to define deployment processes. 

There are some fundamental problems with the way the current `octopusdeploy_deployment_process` resource is built (in combination with historical things about the way Deployment Processes and Step Templates work in the Server API) which mean that although we'd love to, we don't have a sensible way to support consuming Step Templates in deployment processes right now. It's not that we don't want to or are ignoring it, it's that it requires a fundamental rethink of the way Processes are currently represented in Terraform.

We're working on a v2 of the deployment process resource right now, which will alleviate lots of the current problems with Deployment Processes, state drift and step ordering in TF. The new design also unlocks the ability for us to provide a good Step Template consumption experience - it should actually be _better_ than the experience you get via the API directly, mirroring how the UI experience works.  _But_ we know that many of our customers are using the current resource extensively, so we're not taking this lightly. We need to move carefully and ensure it's going to create a real "step"-change (😉) difference for you all, and also make sure that we've got a great migration story in place to minimise the burden of moving to the new resource. 

Thanks to everyone for your (years of) patience with this situation. We're feeling excited about the upcoming new Process resource and the quality of experience it will provide for our Terraform users.

**Comment by mjhilton** [comment]
We're releasing consumption of Step Templates in v1.0 of the Terraform Provider, which will be early next week.

See [this morning's webinar](https://youtu.be/IbCb7x0ckVY?t=1339) for more info on all the great improvements we've put into v1. 

TLDR:
* There is a new Process resource (actually a set of resources) that fixes all the problems with state drift and unexpected planned changes, which replaces the existing DeploymentProcess and RunbookProcess resources
* This new resource also allows us to support Step Template consumption in processes
* You'll need to migrate to the new resource(s), and once you do you can start consuming Step Templates in a first-class way
* We're publishing a migration guide to help you do this, if you already have a deployment process
* As part of this release, we're also migrating this repo and the provider from `OctopusDeployLabs` -> `OctopusDeploy`. The new repo is [here](https://github.com/OctopusDeploy/terraform-provider-octopusdeploy). You'll also need to migrate the provider config name in your HCL from `OctopusDeployLabs` -> `OctopusDeploy` (because Terraform tightly couples repo name to provider name). There'll be a migration guide for that too.

A few more sleeps before we can finally close this issue :)

**Comment by domenicsim1** [comment]
Hi,
We will be closing this issue as its been resolved with the new process resource. Read more [here](https://registry.terraform.io/providers/OctopusDeploy/octopusdeploy/1.0.0/docs/resources/process_templated_step). 

Note that this resource is in the v1.0.0 of the provider, you will need to move to the new official Terraform [namespace](https://registry.terraform.io/providers/OctopusDeploy/octopusdeploy/latest/docs/guides/moving-from-octopus-deploy-labs-namespace) to use it.

Please open future issue on the new GitHub repo [here](https://github.com/OctopusDeploy/terraform-provider-octopusdeploy).
